src/systemctl/systemctl-edit.c

```diff
@@ -483,17 +483,13 @@ static int trim_edit_markers(const char *path) {
                 strshorten(contents_start, contents_end - contents_start);
 
         contents_start = strstrip(contents_start);
-        if (*contents_start && !endswith(contents_start, "\n")) {
-                char *tmp = contents_start;
-                if (MALLOC_SIZEOF_SAFE(contents) - (contents_start - contents) - strlen(contents_start) < 2) {
-                        if ((tmp = realloc(contents, size + 1))) {
-                                contents_start = tmp + (contents_start - contents);
-                                contents = tmp;
-                        }
-                }
-
-                if (tmp)
+        if (!isempty(contents_start) && !endswith(contents_start, "\n")) {
+                char *tmp = contents;
+                if (GREEDY_REALLOC(tmp, (contents_start - contents) + strlen(contents_start) + 2)) {
```

Comment:
(a) yes it is, since we need to maintain the sub-object pointer (`contents_start` is a member of `contents`); and (b) yes we can, greedy_realloc() has normal realloc() semantics (if it fails the old pointer is unchanged), so we just don't add the new-line then (this is a doubly-exceptional state (the user removed the EDIT_MARKER_END *and* the file doesn't end with a new-line *and* the file's size ends up on an allocation boundary *and* we can't allocate more) and it's preferable to generate a sub-optimal file than crash IMO).